src/v/redpanda/admin/services/shadow_link/converter.cc

```diff
@@ -0,0 +1,266 @@
+/// \throws std::invalid_argument if the bootstrap servers are not valid
+cluster_link::model::connection_config
+create_connection_config(const create_shadow_link_request& req) {
+    cluster_link::model::connection_config config;
+    const auto& client_options
+      = req.get_shadow_link().get_configurations().get_client_options();
+    const auto& bootstrap_servers = client_options.get_bootstrap_servers();
+    if (bootstrap_servers.empty()) {
+        throw std::invalid_argument(
+          "bootstrap_servers must not be empty in the client options");
+    }
+    std::ranges::transform(
+      bootstrap_servers,
+      std::back_inserter(config.bootstrap_servers),
+      [](const auto& b) { return net::unresolved_address::from_string(b); });
+
+    if (client_options.has_authentication_configuration()) {
+        config.authn_config = create_authn_settings(
+          client_options.get_authentication_configuration());
+    }
+
+    if (client_options.has_tls_settings()) {
+        set_tls_settings(config, client_options.get_tls_settings());
+    }
+
+    if (client_options.get_metadata_max_age_ms() != 0) {
+        config.metadata_max_age_ms = client_options.get_metadata_max_age_ms();
+    }
+
+    if (client_options.get_connection_timeout_ms() != 0) {
+        config.connection_timeout_ms
+          = client_options.get_connection_timeout_ms();
+    }
+
+    if (client_options.get_retry_backoff_ms() != 0) {
+        config.retry_backoff_ms = client_options.get_retry_backoff_ms();
+    }
+
+    if (client_options.get_fetch_wait_max_ms() != 0) {
+        config.fetch_wait_max_ms = client_options.get_fetch_wait_max_ms();
+    }
+
+    if (client_options.get_fetch_min_bytes() != 0) {
+        config.fetch_min_bytes = client_options.get_fetch_min_bytes();
+    }
+
+    if (client_options.get_fetch_max_bytes() != 0) {
+        config.fetch_max_bytes = client_options.get_fetch_max_bytes();
+    }
+
+    return config;
+}
+} // namespace
+cluster_link::model::metadata
+convert_create_to_metadata(create_shadow_link_request req) {
+    cluster_link::model::metadata metadata;
+
+    try {
+        auto& cluster_link = req.get_shadow_link();
+
+        metadata.name = cluster_link::model::name_t{
+          std::move(cluster_link.get_name())};
+        metadata.uuid = cluster_link::model::uuid_t(uuid_t::create());
+        metadata.connection = create_connection_config(req);
+        metadata.configuration = create_link_configuration(req);
+
+        set_client_id(metadata);
+        return metadata;
+    } catch (const serde::pb::rpc::base_exception& e) {
+        throw;
+    } catch (const std::invalid_argument& e) {
+        throw serde::pb::rpc::invalid_argument_exception(
+          ssx::sformat("Invalid cluster link configuration: {}", e.what()));
+    } catch (const std::exception& e) {
+        throw serde::pb::rpc::unknown_exception(
+          ssx::sformat(
+            "Failed to convert create cluster link request: {}", e.what()));
```

Comment:
should we just let these bubble up? Then the handler can convert this into an `internal` error.